### ISSUE TITLE
Empty reports due to incompatibility with pytest-parallel

### ISSUE THREAD

**Issue by rassie** [pytest-dev/pytest-html#372]
Since 3.0.0 I'm getting empty reports, as in "0 tests ran in 123 seconds". With a bit of bisecting I found out that f2d41db1056a34af78822c2e45c525f7c1e7da67 apparently broke it. Looking into it a bit more, I've found that using `pytest-parallel` with more than one worker makes the individual reports disappear. I assume that post-processing HTML is not really concurrent.

Could someone have a look at that? Thanks.

**Comment by gnikonorov** [comment]
Thank you for raising this @rassie! I'll try and take a look as soon as I can. In the meantime others are more than welcome to fix this. 

Does rolling back to the last release work for you in the interim to fix this issue?

**Comment by rassie** [comment]
Not really, since I need the fix for screenshot embedding, which is also in 3.0.0. But I can manage, since running with one worker is currently a non-issue for me. Thanks for looking into this!

**Comment by gnikonorov** [comment]
> Not really, since I need the fix for screenshot embedding, which is also in 3.0.0. But I can manage, since running with one worker is currently a non-issue for me. Thanks for looking into this!

Of course!

I probably won't be able to get to this for at least a few days.  @ssbarnea @BeyondEvil @csm10495 do any of you have a chance to take a look at this?

**Comment by csm10495** [comment]
I may be able to help. @rassie can you upload a sample simple .py file along with params you are passing to pytest to see something weird happen.
I ran this file (test.py):

```
def testa():
    print('a')

def testb():
    print('b')

def testc():
    print('c')

def testaa():
    print('aa')

def testbb():
    print('bb')

def testcc():
    print('cc')
```

via these args:

```
cmachalo@cmachalo-mn1:~/personal/testing_pytest $ python3 -m pytest test.py --html=out.html --workers 2
==================================================== test session starts =====================================================
platform darwin -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /Users/cmachalo/personal/testing_pytest
plugins: parallel-0.1.0, metadata-1.10.0, cov-2.10.0, html-3.0.0
collected 6 items
pytest-parallel: 2 workers (processes), 1 test per worker (thread)
............
------------------------ generated html file: file:///Users/cmachalo/personal/testing_pytest/out.html ------------------------
===================================================== 6 passed in 0.05s ======================================================

```

and saw this: 

![image](https://user-images.githubusercontent.com/5749838/99344164-6d8da100-285d-11eb-810b-f6d7ed42192c.png)


which looked ok. Also does the issue happen every time?

**Comment by gnikonorov** [comment]
Thank you @csm10495 !

**Comment by csm10495** [comment]
I did notice when running it over and over sometimes logging was in the wrong place, though that can also be reproduced using junitxml (and is therefore not a pytest-html specific bug). Using the same test.py as above.

I ran: 

```
cmachalo@cmachalo-mn1:~/personal/testing_pytest 4 $ python3 -m pytest test.py --junitxml=out.xml --workers 2 -o junit_logging=all
==================================================== test session starts =====================================================
platform darwin -- Python 3.8.3, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/cmachalo/personal/testing_pytest
plugins: parallel-0.1.0, metadata-1.10.0, html-2.0.2.dev57+g67bd83d.d20201117, cov-2.10.0, xdist-2.1.0, forked-1.3.0
collected 6 items
pytest-parallel: 2 workers (processes), 1 test per worker (thread)
............
---------------------------- generated xml file: /Users/cmachalo/personal/testing_pytest/out.xml -----------------------------
===================================================== 6 passed in 0.05s ======================================================
```

junit generated:
```
<testsuites>
<testsuite name="pytest" errors="0" failures="0" skipped="0" tests="6" time="0.055" timestamp="2020-11-16T23:52:02.243555" hostname="cmachalo-mn1.linkedin.biz">
<testcase classname="test" name="testa" time="0.002">
<system-out>--------------------------------- Captured Log --------------------------------- --------------------------------- Captured Out --------------------------------- a </system-out>
<system-err>--------------------------------- Captured Err --------------------------------- </system-err>
</testcase>
<testcase classname="test" name="testb" time="0.002">
<system-out>--------------------------------- Captured Log --------------------------------- --------------------------------- Captured Out --------------------------------- b </system-out>
<system-err>--------------------------------- Captured Err --------------------------------- </system-err>
</testcase>
<testcase classname="test" name="testc" time="0.001">
<system-out>--------------------------------- Captured Log --------------------------------- --------------------------------- Captured Out --------------------------------- c </system-out>
<system-err>--------------------------------- Captured Err --------------------------------- </system-err>
</testcase>
<testcase classname="test" name="testaa" time="0.001">
<system-out>--------------------------------- Captured Log --------------------------------- --------------------------------- Captured Out --------------------------------- aa </system-out>
<system-err>--------------------------------- Captured Err --------------------------------- </system-err>
</testcase>
<testcase classname="test" name="testbb" time="0.001">
<system-out>--------------------------------- Captured Log --------------------------------- --------------------------------- Captured Out --------------------------------- </system-out>
<system-err>--------------------------------- Captured Err --------------------------------- </system-err>
</testcase>
<testcase classname="test" name="testcc" time="0.001">
<system-out>--------------------------------- Captured Log --------------------------------- --------------------------------- Captured Out --------------------------------- bb cc </system-out>
<system-err>--------------------------------- Captured Err --------------------------------- </system-err>
</testcase>
</testsuite>
</testsuites>
```
Notice how testcc has bb and cc instead of just cc. testbb has empty logging. If this is the issue that we're seeing, we should move this to pytest-parallel as it isn't specific to pytest-html

**Comment by rassie** [comment]
I can reproduce the problem with this simple test: 

```
def pytest_generate_tests(metafunc):
    if "number" in metafunc.fixturenames:
        numbers = range(0, 50)
        metafunc.parametrize("number", numbers)


def test_numbers(number):
    assert number % 2 == 0
```

Running with `poetry run python -m pytest --workers 2 tests/ --html=tests/reports/index.html`, I get no results in the HTML output. Setting `--workers 1` returns full report.

**Comment by rassie** [comment]
Adding `--junitxml=out.xml` with two workers produces a mostly empty `out.xml`, so the problem might indeed not be `pytest-html`-specific.

**Comment by rassie** [comment]
I'll try upgrading my dependencies first. Thanks for the help!

**Comment by rassie** [comment]
I can confirm the problem is solved with `pytest-parallel` 0.1.0 (instead of 0.0.10). However, it might be interesting to know why `pytest-html` 2.1.1 worked with that older `pytest-parallel` version, while 3.0.0 does not. But either way, thank you everyone, this particular problem is solved.

**Comment by ssbarnea** [comment]
TBH, I will not be able to work on any bugs this year but I will do my best to review PRs. As and old guru said at some point: in open-source, you need it, you fix it. ;)
